google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -100,10 +91,11 @@ void TracingMessageBatch::Flush() {
 
   // This must be called before we clear the message spans.
   AddMessageSpanMetadata();
-  // Clear message spans.
-  message_spans_.clear();
-
-  batch_sink_spans_.push_back(batch_sink_parent_span_);
+  {
+    std::lock_guard<std::mutex> lk(mu_);
```

Comment:
Feel free to move this suggestion to a different PR.

Now the code is grabbing and releasing the lock 3 times in quick succession.  That suggests the critical section(s) are not organized really well.

Maybe something like this instead?

```cc
std::unique_lock<std::mutex> lk(mu_);
auto parent_span = MakeParentSpan(std::move(message_spans_));
message_spans_.clear();
batch_sink_spans_.push_back(parent_span);
lk.unlock();

auto async_scope = internal::GetTracer(internal::CurrentOptions())->WithActiveSpan(parent_span);
child->Flush();
```

where

```cc
auto MakeParentSpan(std::vector<...> message_spans) {
  std::vector<std::pair<SpanContext, AttributesList>> links(message_spans.size());
  std::transform(message_spans.begin(), message_spans.end(), links.begin(), // yes, not back_inserter()
         [](...) { .... });
   auto parent_span = internal::MakeSpan("BatchSink::AsyncPublish",
                           /*attributes=*/
                           {{"messaging.pubsub.num_messages_in_batch",
                             static_cast<std::int64_t>(message_spans.size())}},
                           /*links*/ links);
  auto context = parent_span->GetContext();
  auto trace_id = internal::ToString(context.trace_id());
  auto span_id = internal::ToString(context.span_id());
  for (auto& m : message_spans) {
    m->AddEvent("gl-cpp.batch_flushed");
    m->SetAttribute("pubsub.batch_sink.trace_id", trace_id);
    m->SetAttribute("pubsub.batch_sink.span_id", span_id);
  }
  return parent_span;
}
```

Reply:
Rewrote so the critical section is longer.